### ISSUE TITLE
Support more audio and video file formats by integrating LibVLC or ffmpeg

### ISSUE THREAD

**Issue by achimmihca** [UltraStar-Deluxe/Play#362]
## Actual behaviour

Unity supports a limited, platform dependent set of audio and video file formats.
This is causing issues with existing UltraStar songs and is a recurring issue for user discussions (#344, #246).

Plus, Unity has a bug when decoding mp3 with variable bitrate (#323).

## Expected behaviour

By integrating a third party lib like LibVLC or ffmpeg, it should be possible to support the plethora of file formats that users expect.

## ffmpeg
There are C# bindings for ffmpeg:
- https://github.com/rosenbjerg/FFMpegCore
- https://github.com/Ruslan-B/FFmpeg.AutoGen
- [more on GitHub](https://github.com/search?q=ffmpeg+c%23)

and some attempts to integrate it in Unity:
- https://github.com/ViveSoftware/ViveMediaDecoder
- [more on GitHub](https://github.com/search?p=4&q=ffmpeg+unity&type=Repositories)

However, most of these attempts to integrate ffmpeg with Unity seem like an early proof-of-concept to me. I did not find anything polished for desktop and nothing for Android or iOS.

## LibVLC
There are C# bindings for LibVLC:
- https://github.com/videolan/libvlcsharp

and an integration for Unity:
- https://code.videolan.org/videolan/vlc-unity
    - This project seems active, with commercial support
    - So far, only Windows and Android are supported

The [LibVLC wiki](https://wiki.videolan.org/LibVLC) names all important platforms for UltraStar Play:

> Run on every platform, from desktop (Windows, Linux, Mac) to mobile (Android, iOS) and TVs (AppleTV, Android TV, Tizen) and Consoles (PS4/5, Xbox)

Plus, LibVLC seems to support network file systems, which could solve #347

> Network browsing for distant filesystems (SMB, FTP, SFTP, NFS...) and servers (UPnP, DLNA)

In theory, it should be possible to make LibVLCSharp and its Unity integration support the same features.

## License considerations

LibVLC and ffmpeg are both licensed under [LGPL 2.1](https://www.gnu.org/licenses/lgpl-3.0.html) or later.
From my understanding, this is OK for us (at least on desktop platforms) because Unity stores the libs as dynamically linked library (DLL) in the app directory (in Windows they are in `./UltraStar Play_Data/Managed`, similar for macOS and Linux, not sure about iOS and Android though).
Thus, users are able to swap the implementation of the bundled libs.

From my understanding, this allows us to distribute the rest of the app (or "combined work" as the license calls it) "under terms of your choice", in our case under MIT license.

### Closed source license considerations
For closed source projects, [ffmpeg](https://ffmpeg.org/legal.html) does not provide any other licensing.

> Note that FFmpeg is not available under any other licensing terms, especially not proprietary/commercial ones, not even in exchange for payment. 

In contrast, LibVLCSharp is also available under [paid commercial license](https://videolabs.io/solutions/libvlcsharp/) by videolabs.

## Conclusion

The Unity integration of LibVLC seems more mature and the option for a commercial license offers more flexibility also for closed source projects.
Further, it seems LibVLC has been developed with all relevant platforms in mind and its distant filesystems feature could be handy as well.

Thus, I will try to integrate LibVLC for UltraStar Play.

**Comment by achimmihca** [comment]
The free trial version of [LibVLC for Unity](https://videolabs.io/solutions/vlc-unity-trial.unitypackage) works like a charm on Windows.

Question is how to make it run on Linux and macOS.

**Comment by basisbit** [comment]
The custom ported mp3 parsing which was integrated in US Play previously supported variable bitrates without issues. I'd suggest going back to that. Regarding LibVLC, I'd argue that you can not legally sell a closed source version of the result, and I'd argue that the integration plugin in the unity packages store is also violating LGPL.

Also, this approach does not help for the majority of potential users: mobile device users. Instead, it makes it less likely that those who publish song packages adjust their tooling so that mibile devices also work with it. I'd suggest focusing on the song packages and song repositories idea instead (just my 2ct)

**Comment by achimmihca** [comment]
### About license considerations

I asked on the [Unity forums](https://forum.unity.com/threads/are-lgpl-licensed-libraries-compatible-with-unity-build-output.1384116/) and [VLC for Unity GitLab](https://code.videolan.org/videolan/vlc-unity/-/issues/174) whether mixing LGPL libs with an app that uses another license is possible in general on Windows / Linux / macOS / iOS / Android.

It seems to be possible in general because by now, all of these platforms support dynamic linking.
Furthermore, there should be plenty of example apps that are not under LGPL themselves but use ffmpeg, libvlc, or Qt.

One needs to verify this and provide instructions how to swap a lib in an app built with Unity on Android / iOS / tvOS. But at least it seems to be possible *somehow*.

**Comment by achimmihca** [comment]
> One needs to verify this and provide instructions how to swap a lib in an app built with Unity on Android

- Get apk file from Android device using adb ( https://stackoverflow.com/questions/4032960/how-do-i-get-an-apk-file-from-an-android-device )
    - Determine the package name of the app: `adb shell pm list packages`
    - Get the full path name of the APK file for the desired package: `adb shell pm path com.example.someapp`
    - Using the full path name from Step 2, pull the APK file from the Android device to the development box: `adb pull /data/app/com.example.someapp-2.apk path/to/desired/destination`
- Unpack the apk, swap lib, pack the apk using [apktool](https://ibotpeaches.github.io/Apktool/)
   - unpack: `apktool d test.apk`
   - swap lib. The DLLs can be found in the extracted apk in `assets\bin\Data\Managed` 
   - pack: `apktool b test`

This process allows to swap a dynamically linked lib. Thus, I think using an LGPL lib on Android is not an issue.

---
During my research I found that Android and its ecosystem are open, transparent, and versatile. I appreciate the platform even more than before.

And on the other side, there is Apple... closed, intransparent, and restricted by arbitrary regulations.

I can't even upload an app to my device. Apple does not allow to install current macOS on my macbook, does not allow to install XCode on the old macOS version, does not allow to build and upload the app to iOS 16 with old XCode.
WTF?! Even if the XCode version can only build for iOS 15, is the iOS 16 device incompatible so badly? Like, every new iOS version requries a rebuild of every app there is?

Probably not, there are projects to leverage macOS restrictions (https://github.com/dortania/OpenCore-Legacy-Patcher), which makes me assume that Apple just wants to sell new stuff by focing existing stuff into obsolescence.
Easiest way to solve this is to buy a more recent device, which - oh surprise - costs at least twice as much.

In conclusion, iOS may take some more time until I get to it.
Sorry for the rant.

**Comment by achimmihca** [comment]
With respect to iOS, one can

- Get the installed [ipa file from the device](https://medium.com/@contact.jmeyers/download-the-ipa-file-for-any-ios-app-on-your-iphone-8298c7d6d812) using Apple Configurator
- unzip the ipa file. This yields a folder "Payload" that contains the .app folder

But then you will find that the individual C# DLLs are not present in the .app folder. There is a folder "Data/Managed", which is similar to the apk. But in contrast to the apk, the ipa conterpart does not contain individual libs.

This was expected.
So, I think this is an issue for using an LGPL lib in a **closed source** iOS app.

However, iOS seems to support some dynamic linking mechanism (dylib files).
Question is whether this can be used with Unity. I don't know but I will not dig deeper.

**Comment by achimmihca** [comment]
I wonder whether using a LGPL lib in an **MIT licensed** project is an issue in the first place.

Actually, I don't think so because the MIT licensed project offers the source code and everything to relink and rebuild the application. I think this satisfies the LGPL.

LGPL 4d0:
> 0) Convey the Minimal Corresponding Source under the terms of this License, and the Corresponding Application Code in a form suitable for, and under terms that permit, the user to recombine or relink the Application with a modified version of the Linked Version to produce a modified Combined Work, in the manner specified by section 6 of the GNU GPL for conveying Corresponding Source.

As I understand it
- we (as MIT licensed project) include the LGPL licensed lib as-is ("under the terms of this License")
- and we convey "the Corresponding Application Code in a form suitable for, and under terms that permit, the user to recombine or relink the Application with a modified version of the Linked Version to produce a modified Combined Work..."

I just don't understand the part "in the manner specified by section 6 of the GNU GPL for conveying Corresponding Source". Section 6 of GPL is about "Conveying Non-Source Forms". But an MIT licensed project even conveys the "source form".

Only issue I see is **everything we include under the terms of Unity**. I guess this might be an issue.

Phew! I am not a lawyer and English is not my native language so I am lost ;)
As it is now, I would not include libVLC in an iOS build.

Any further thoughts on the topic?

**Comment by achimmihca** [comment]
> Thus, I think using an LGPL lib on Android is not an issue

Except it can still conflict with additional terms of app stores and distribution channels.
More specifically, the [Google Play terms](https://play.google.com/about/play-terms) do not allow modifying any content you've obtained through the store. You can see this clause under Section 4 (Rights and Restrictions).

Thanks for pointing this out on the [Unity forums](https://forum.unity.com/threads/are-lgpl-licensed-libraries-compatible-with-unity-build-output.1384116/#post-8813605)

So, I am back where I started.
LGPL is most likely not an issue on Desktop platforms but it most likely is an issue on mobile platforms.

---
I am unsure what to do.

Better not include libVLC in the first place?

The supported media file formats are already platform specific. libVLC could at least unify this for desktop platforms.

**Comment by basisbit** [comment]
Thank you so much for the work to clear up the legal side. My personal opinion / suggestion would be, to use LibVLC only for the desktop-targeting platforms for now. That should already alleviate a majority of difficulty for the average users. Having h264 on mp4 support for all platforms gives us universal support for the most widely used format among the usdx files creating community. This also gives wider formats support for Android targets that are not Google-walled, including various "cheap" TV connectors/boxes.

**Comment by jimmyhawkin** [comment]
SOunds awsome tha towrk is being done on this. I was just about to post about it. But im glad to see that support for this is being worked on.

**Comment by g3n35i5** [comment]
@achimmihca I'd love to see this feature included in Play! Is there any way to help you implementing it?

**Comment by achimmihca** [comment]
> Is there any way to help you implementing it

What is needed is a good integration of ffmpeg or libVLC in Unity.

Some thoughts so far
- Both ffmpeg and libVLC have Unity plugins on the asset store. These cannot be used for UltraStar Play directly, but their approach could be re-implemented for UltraStar Play with a fitting license.
    - ffmpeg: https://assetstore.unity.com/packages/tools/video/ffmpeg-for-unity-199811
    - libVLC: https://assetstore.unity.com/packages/tools/video/vlc-for-unity-windows-133979
    - both of these have bad reviews, libVLC is expensive, ffmpeg unstable.
    - libVLC on the Asset Store is not available for all targeted platforms (Win,Linux,macOS,iOS,Android).
        - Seems to require a lot of work to make libVLC working on other platforms, see [VLC for Unity GitLab](https://code.videolan.org/videolan/vlc-unity/-/issues/19)
    - ffmpeg on the Asset Store is cheap, includes the set of ffmpeg libs with proper license, and should work on all targeted platforms (Win,Linux,macOS,iOS,Android). This is a very good starting point.
        - However, ffmpeg on the Asset Store is unstable. A better implementation or at least some fixes of the plugin would be needed.
- GitHub has few repos that try to integrate libVLC in Unity:
    - https://github.com/search?q=vlc%20unity&type=repositories
    - https://github.com/bosqmode/UnityVLCPlayer
- GitHub has some more repos that try to integrate ffmpeg in Unity:
    - https://github.com/search?q=ffmpeg+unity&type=repositories
    - FFmpegUnityBind2 seems to be the most complete and stable https://github.com/Spirit30/FFmpegUnityBind2
        - It can call ffmpeg commands.
    - However, these are typically incomplete and missing direct video / audio playback.
- GitHub has some ffmpeg bindings in C#: https://github.com/Ruslan-B/FFmpeg.AutoGen
- Both ffmpeg and libVLC would work with respect to their license on Windows/macOS/Linux. Expect some libs that are GPL, these must not be included in the build.
    - Android and iOS may be a problem with LGPL, see my thoughts on this above

I integrated ffmpeg for Unity in Melody Mania. It nearly works but it is unstable.
I will try libVLC for Unity next, but it is only available for Windows, which is not a solution in the long run.

Thus, what I would try to do is
- take the ffmpeg plugin for Unity and learn from it
    - it already has the required libs and [good documentation]() to [build ffmpeg](https://github.com/NON906/ffmpeg-windows-build-helpers)
    - although unstable, the approach seems to work in general.
    - You could start with `FfmpegPlayerImpWinMonoBin`
- take FFmpegUnityBind2 or FFmpeg.AutoGen as alternative to the Asset Store plugin to run ffmpeg commands
- try to decode audio/video via ffmpeg
    - maybe this can be achieved with commands (decode to raw frame and raw samples)?
        - Something like `-f rawvideo -pix_fmt rgba`
        - this seems to be done by the ffmpeg Unity plugin on the asset store.
- load decoded audio samples and video pixels in Unity
    - pixels of the video could be loaded in a texture: Texture2D.LoadRawTextureData
    - samples of the audio could be played in [OnAudioFilterRead](https://github.com/TriggerCoder/CSharpSynthForUnity2.0/blob/main/Assets/Scripts/PcmReaderCallbackTest.c

**Comment by g3n35i5** [comment]
> @g3n35i5 , see above "What I would like to see"

Thanks for your answer! I hope that in the near future I will find time to deal with it.

**Comment by achimmihca** [comment]
There are ffmpeg integrations for other game engines such as Godot
- https://github.com/kidrigger/godot-videodecoder
- https://github.com/Godot-dcl/ffmpeg-node
- https://github.com/DataPlusProgram/gdVideoPlayer

Maybe their approach can also be used to integrate ffmpeg in Unity.
